File: week5/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int longestPath = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        maxDepth(root);
        return longestPath;
    }
    
    public int maxDepth(TreeNode root){
        if(root == null) {
            return 0;
        }
        // Get the maximum depth of the left side
        int left = maxDepth(root.left);
        // Get the maximum depth of the right side
        int right = maxDepth(root.right);
        // Figure out which is bigger - the current max or the path
        // from the left most to right most
        longestPath = Math.max(left + right, longestPath);
        // Return the depth of the root
        return Math.max(left, right) + 1;
    }
}

File: summer2021/unit5/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        result = 0
        if not root:
            return result
        leftSide = 0
        if root.left:
            leftSide = self.height(root.left)
        rightSide = 0
        if root.right:
            rightSide = self.height(root.right)
        return max(leftSide + rightSide, max(self.diameterOfBinaryTree(root.left), self.diameterOfBinaryTree(root.right)))
        
    def height(self, root):
        if not root:
            return 0
        return 1 + max(self.height(root.left), self.height(root.right))